File: JavaScript/902-Numbers At Most N Given Digit Set.js
/**
 * @param {string[]} digits
 * @param {number} n
 * @return {number}
 */
var atMostNGivenDigitSet = function(digits, n) {
    let res = 0, strN = n.toString();

    for (let i=1; i<strN.length; i++){
        res += Math.pow(digits.length, i);
    }
    
    for (let i=0; i<strN.length; i++){
        let matchPrefix = false;
        for (let d of digits){
            if (parseInt(d) < parseInt(strN[i])){
                res += Math.pow(digits.length, strN.length-i-1);
            }else if (parseInt(d) === parseInt(strN[i])){
                matchPrefix = true;
                break;
            }
        }
        if (!matchPrefix) return res;
    }
    return res + 1;
};

File: 902-Numbers At Most N Given Digit Set.py
class Solution(object):
    def atMostNGivenDigitSet(self, digits, n):
        """
        :type digits: List[str]
        :type n: int
        :rtype: int
        """
        res = 0
        
        strN = str(n)
        N = len(strN)
        
        for i in range(1, N):
            res += pow(len(digits), i)
            
        for i in range(N):
            prefixMatch = False
            for d in digits:
                if int(d) < int(strN[i]):
                    res += pow(len(digits), N-i-1)
                elif int(d) == int(strN[i]):
                    prefixMatch = True
                    break
            if not prefixMatch:
                return res
                    
        return res+1